### PR TITLE
Revamped version of CotterAPIService

### PR DESCRIPTION
**Pull request by albertpurnama**

### CODE REVIEW

**Comment by albertpurnama** [comment]
https://app.clickup.com/t/29ndxh

File: Source/Service/Requests/CreateAuthenticationEvent.swift [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+//
+
+import Foundation
+
+public struct CreateAuthenticationEvent: APIRequest {
+    public typealias Response = CotterEvent
+    
+    public var path: String {
+        return  "/event/create"
+    }
```
Comment (calvinnfernando) [inline Source/Service/Requests/CreateAuthenticationEvent.swift:15]:
```suggestion
    public let path: String = "/event/create"
```

File: Source/Service/Requests/CreateAuthenticationEvent.swift [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    public var method: String {
+        return "POST"
+    }
```
Comment (calvinnfernando) [inline Source/Service/Requests/CreateAuthenticationEvent.swift]: same

File: Source/Service/Requests/EnrollUserPIN.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+//
+
+import Foundation
+
+public struct EnrollUserPIN: APIRequest {
+    public typealias Response = CotterUser
+    
+    public var path: String {
+        return "/user/" + self.userID
+    }
```
Comment (calvinnfernando) [inline Source/Service/Requests/EnrollUserPIN.swift]: same, in addition: `"/user/\(self.userID)"`

^this feature is called string interpolation and generally we should prefer interpolation over concatenation in many diff languages, [here](https://teamtreehouse.com/community/what-is-the-difference-between-concatenation-and-interpolation-it-seems-like-you-can-do-the-exact-same-thing-with-both) is (a not so short reading) why. does not really apply here but just for future reference.

File: Source/Service/Requests/EnrollUserPIN.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    
+    public var method: String {
+        return "PUT"
+    }
```
Comment (calvinnfernando) [inline Source/Service/Requests/EnrollUserPIN.swift]: same

File: Source/Service/Requests/GetBiometricStatus.swift [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        return "/user/enrolled/" + self.userID + "/BIOMETRIC/" + self.pubKey
+    }
+    
+    public var method: String {
+        return "GET"
+    }
+    
+    public var body: Data? {
+        return nil
+    }
```
Comment (calvinnfernando) [inline Source/Service/Requests/GetBiometricStatus.swift]: same

File: Source/Service/Requests/GetUser.swift [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    }
+    
+    public var method: String {
+        return "GET"
+    }
+    
+    public var body: Data? {
+        return nil
+    }
+
```
Comment (calvinnfernando) [inline Source/Service/Requests/GetUser.swift]: again

File: Source/Service/Requests/RegisterBiometric.swift [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        let data: [String:Any] = [
+            "method": "BIOMETRIC",
+            "enrolled": true,
+            "code": pubKey
+        ]
+        
+        let body = try? JSONSerialization.data(withJSONObject: data)
+        
+        return body
+    }
```
Comment (calvinnfernando) [inline Source/Service/Requests/RegisterBiometric.swift]: same

File: Source/Service/Requests/RegisterUser.swift [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+public struct RegisterUser: APIRequest {
+    public typealias Response = CotterUser
+    
+    public var path: String {
+        return "/user/create"
+    }
+    
+    public var method: String {
+        return "POST"
+    }
```
Comment (calvinnfernando) [inline Source/Service/Requests/RegisterUser.swift]: same

File: Source/Service/Requests/RequestToken.swift [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    public typealias Response = CotterIdentity
+
+    public var path: String {
+        return "/verify/get_identity"
+    }
+
+    public var method: String {
+        return "POST"
+    }
+
```
Comment (calvinnfernando) [inline Source/Service/Requests/RequestToken.swift]: same

File: Source/Service/Requests/UpdateBiometricStatus.swift [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+public struct UpdateBiometricStatus: APIRequest {
+    public typealias Response = CotterUser
+    
+    public var path: String {
+        return "/user/" + self.userID
+    }
+    
+    public var method: String {
+        return "PUT"
+    }
```
Comment (calvinnfernando) [inline Source/Service/Requests/UpdateBiometricStatus.swift]: 👀

File: Source/Service/Requests/UpdateUserPIN.swift [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    public typealias Response = CotterUser
+    
+    public var path: String {
+        return "/user/" + self.userID
+    }
+    
+    public var method: String {
+        return "PUT"
+    }
+
```
Comment (calvinnfernando) [inline Source/Service/Requests/UpdateUserPIN.swift]: 👀

**Review by calvinnfernando — COMMENTED** [review]
left nitpicks but i luv it. A1 quality albert you're doing god's work

**Comment by rayandrie** [comment]
Nothing much to add besides Calvin's comments. [LGTM!](https://www.youtube.com/watch?v=rR4n-0KYeKQ)

File: Source/Service/Requests/CreateAuthenticationEvent.swift (hunk 1 above)
Comment (albertpurnama) [reply Source/Service/Requests/CreateAuthenticationEvent.swift:15]: path may have custom parameters that need to be defined inside the init, so using `let` wouldn't be possible.

So I will leave this as is for consistency

File: Source/Service/Requests/CreateAuthenticationEvent.swift (hunk 2 above)
Comment (albertpurnama) [reply Source/Service/Requests/CreateAuthenticationEvent.swift]: I will change this to let since it is constant

**Review by rayandrie — APPROVED** [review]

**Pull request merged**